### PR TITLE
#2220, integrated aider style linting, currently passes related o…

### PR DESCRIPTION
**Pull request by jigsawlabs-student**
WIP for issue #2220 

Currently uses Aider style linting, and passes existing test_agent_skill.py tests.  

Needed to update aider_linter slightly to accomplish this.   Biggest change to aider_linter is in Linter#py_lint function, where do not proceed with all linters, but stop at the first python linter to get an error.  This was done because was catching the same issue twice.  

Todo: 

* Should add additional tests to confirm that will now have basic linting functionality for other languages.  
* Potential refactoring where the Aider style linter detects line number with the LintResult(), instead being done in OpenDevin's _edit_or_append_file function.
* The error format of the linter is currently different depending on which pylinter is used
* Should we run all linters then remove where same line caught?
* Could not find aider based tests associated with Linter, so can add some of these where necessary

### CODE REVIEW

**Comment by tobitege** [comment]
Hello!

Great to see more contributions! I had already started on this Linter integration, but due to other tasks, this got a little behind: https://github.com/OpenDevin/OpenDevin/pull/2271/

In this commit you'll find a unit test file with several basic tests: https://github.com/OpenDevin/OpenDevin/pull/2271/commits/11f8a457546415925193ad90b495018ea0567806
Feel free to copy it over if they're usefull or save you time.

I'm happy to close my PR in favor of yours, looking forward to your progress! 👍

**Comment by tobitege** [comment]
Sorry, clicked in the wrong tab.

**Comment by jigsawlabs-student** [comment]
@tobitege ok excellent, thanks for referencing your commit, didn't realize that, but will take a look/integrate.  I'll put more work into this later in the week.

File: opendevin/runtime/plugins/agent_skills/agentskills.py
Comment (tobitege) [inline opendevin/runtime/plugins/agent_skills/agentskills.py:865]: This is exactly what OpenDevin would do 😬

File: tests/unit/test_sandbox.py
Comment (JeffKatzy) [inline tests/unit/test_sandbox.py:321]: @tobitege I do think there is a legitimate bug my code is somehow leading to.  For some reason, the output with the pr is `/workspace/hello.py` instead of `/workspace/test/hello.py`.  Not sure what's going on given that only the `_lint_file()` function is changed, but will look into it.  If anyone has initial ideas, would appreciate it.

Comment (tobitege) [reply tests/unit/test_sandbox.py:321]: First, please make sure you ran "make lint" and see if there are open issues to fix. ;)

File: poetry.lock
Comment (tobitege) [inline poetry.lock:1376]: Merge errors

File: pyproject.toml
Comment (tobitege) [inline pyproject.toml:66]: is this required instead of "*"?

File: tests/unit/test_sandbox.py
Comment (tobitege) [reply tests/unit/test_sandbox.py:321]: this path issues looks like you're running with that persist... option being true.
switch to false instead and try again.

File: tests/unit/test_aider_linter.py
Comment (tobitege) [inline tests/unit/test_aider_linter.py:48]: nice to seem some other languages, too! 👍

**Review by tobitege — CHANGES_REQUESTED** [review]
Overall looks good, good job!
Only minor fixes to tweak (e.g. poetry lock file is bust) and make lint needed.
Reach out when we shall take another look!

File: tests/unit/test_aider_linter.py
Comment (JeffKatzy) [reply tests/unit/test_aider_linter.py:48]: @tobitege Appreciate that! Ok, so the test_sandbox, make lint updates, and poetry.lock should be taken care of.

**Comment by tobitege** [comment]
If ready for review, go and click the Ready for review button. 😏

**Comment by JeffKatzy** [comment]
@tobitege Ready for review 

Locally, I do get a failing test, but I also see this test failing on main

```
if default_config.llm.api_key is not None:  # prevent leak
>           pytest.fail('LLM API key should be empty.')
E           Failed: LLM API key should be empty.

tests/unit/test_config.py::test_invalid_toml_format - Failed: LLM API key should be empty.
```

Cannot click the 'request review' button as the pull request is marked as draft

`Only those with [write access](https://docs.github.com/articles/what-are-the-different-access-permissions) to this repository can mark a draft pull request as ready for review.`

**Comment by tobitege** [comment]
> Locally, I do get a failing test, but I also see this test failing on main
> ```
> tests/unit/test_config.py::test_invalid_toml_format - Failed: LLM API key should be empty.
> ```

The above unit test error is only happening locally, but will not happen on the server as there won't be a key filled.

I'll have a look now! 😀

File: opendevin/runtime/plugins/agent_skills/aider/README.md
Comment (tobitege) [inline opendevin/runtime/plugins/agent_skills/aider/README.md:8]: I think it's "its"? 😂

File: opendevin/runtime/plugins/agent_skills/setup.sh
Comment (tobitege) [inline opendevin/runtime/plugins/agent_skills/setup.sh:21]: hmm why `litellm` here?

**Review by tobitege — CHANGES_REQUESTED** [review]
Overall it looks great!
Just 2 little nits, the litellm is probably from testing?

File: opendevin/runtime/plugins/agent_skills/setup.sh
Comment (tobitege) [reply opendevin/runtime/plugins/agent_skills/setup.sh:21]: Oh, and remove "WIP" from PR title ;)

Comment (JeffKatzy) [reply opendevin/runtime/plugins/agent_skills/setup.sh]: @tobitege Ok, fixed!

**Review by tobitege — APPROVED** [review]

**Comment by tobitege** [comment]
@JeffKatzy Great! Thanks so much!

@xingyaoww @neubig This linter is ready and I'd like you to have the final word, in case you'd like to do an eval run with this branch first etc.

**Comment by tobitege** [comment]
@JeffKatzy sorry for the delay, but could you look into those conflicts?
There have been some bigger refactors, that now need your code to be adapted accordingly.

**Review by xingyaoww — APPROVED** [review]
Looks amazing - thanks for your contribution!
The aider seems to be integrated in a nice & clean way!

Btw, now we `aider` integrated so nicely, does it mean we can get back repo map (https://github.com/OpenDevin/OpenDevin/pull/2248) very easily as well by reusing the same plugin? (cc @ryanhoangt)

**Review by neubig — APPROVED** [review]
Looks awesome, let me quickly resolve the conflicts! They're easy to fix.

**Comment by ryanhoangt** [comment]
> Btw, now we aider integrated so nicely, does it mean we can get back repo map (https://github.com/OpenDevin/OpenDevin/pull/2248) very easily as well by reusing the same plugin? (cc @ryanhoangt)

Hmm, not really. In my previous PR, I also finished integrating it similarly to this. The caveats are that it can't tap into the message history for now and also significantly downgrades performance.

**Review by neubig — CHANGES_REQUESTED** [review]
Just to be sure that this doesn't break anything I'm running a quick swe-bench eval (maybe not the whole thing) to make sure this is working, we can merge after that. I'll do "request changes" just to block merging until I'm done. I don't have any issues with the PR.

**Comment by neubig** [comment]
Hmm, I ran an eval on this and I'm getting lots of errors like `open_file not defined` and `edit_file not defined`:

```
==============
CodeActAgent LEVEL 0 STEP 0

[92m06:51:15 - opendevin:INFO[0m: llm.py:245 - Cost: 0.02 USD | Accumulated Cost: 0.02 USD
06:51:15 - ACTION
**IPythonRunCellAction**
THOUGHT: Thank you for providing the problem statement. Let's investigate this issue and try to fix it. First, I'll check the current implementation of the `separability_matrix` function.
CODE:
open_file('/workspace/astropy__astropy__4.3/astropy/modeling/separable.py')
06:51:22 - OBSERVATION
**IPythonRunCellObservation**
Failed to connect to the Jupyter server <html><title>500: Internal Server Error</title><body>500: Internal Server Error</body></html>
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Cell In[1], line 1
----> 1 open_file('/workspace/astropy__astropy__4.3/astropy/modeling/separable.py')

NameError: name 'open_file' is not defined
```

They seem to be unrelated to this PR, but they're worrisome, it'd be good to check if we get these on main as well.

**Comment by tobitege** [comment]
`Failed to connect to the Jupyter server <html><title>500: Internal Server Error</title><body>500: Internal Server Error</body></html>`

**Comment by xingyaoww** [comment]
@neubig is it related to the latest sandbox upgrade?
It looks like there's something wrong with the jupyter server inside the sandbox..

Maybe the sandbox test should be fixed first to aid the debug process https://github.com/OpenDevin/OpenDevin/actions/runs/9825337018/job/27125574267?pr=2489

**Comment by neubig** [comment]
Hmm, OK. It seems the sandbox test is failing just because the formatting of the message changed? We should fix that of course, but that itself is not the root cause of the problem that I witnessed.

**Comment by xingyaoww** [comment]
@neubig it is possible that there will be some other sandbox related error that get obscured by the first formatting error - i can try to take a look

**Comment by xingyaoww** [comment]
@neubig sandbox tests should passed. I also locally ran a few instances and it seemed ok?

**Review by neubig — APPROVED** [review]
Hi! I'm on vacation this week and can't run the eval, so I'll cede to @xingyaoww 's judgement on this one.

**Comment by xingyaoww** [comment]
@jigsawlabs-student Let me know when you fixed the unit tests and i'd be happy to run an eval on this!

**Comment by JeffKatzy** [comment]
@xingyaoww 

I am unable to reproduce this locally, as it's failing at the install and start docker step, and specifically `sudo ln -sf $HOME/.colima/default/docker.sock /var/run/docker.sock`.

It looks like it did successfully move through moving through steps 1 through 4 of booting up colima.  But then appears to move through those steps again and then fails.  

Any thoughts as to what may be going on?


```
time="2024-07-10T23:01:37Z" level=info msg="QEMU binary \"/usr/local/bin/qemu-system-x86_64\" seems properly signed with the \"com.apple.security.hypervisor\" entitlement"
time="2024-07-10T23:01:38Z" level=info msg="[hostagent] hostagent socket created at /Users/runner/.colima/_lima/colima/ha.sock"
time="2024-07-10T23:01:38Z" level=info msg="[hostagent] Using system firmware (\"/usr/local/share/qemu/edk2-x86_64-code.fd\")"
time="2024-07-10T23:01:38Z" level=info msg="[hostagent] Starting QEMU (hint: to watch the boot progress, see \"/Users/runner/.colima/_lima/colima/serial*.log\")"
time="2024-07-10T23:01:38Z" level=info msg="SSH Local Port: 49245"
time="2024-07-10T23:01:38Z" level=info msg="[hostagent] Waiting for the essential requirement 1 of 4: \"ssh\""
time="2024-07-10T23:01:48Z" level=info msg="[hostagent] Waiting for the essential requirement 1 of 4: \"ssh\""
```

File: poetry.lock
Comment (yufansong) [inline poetry.lock:1]: It seems you update too much unreleated package, please revert them before merge. And only update the necessary dependency.

Comment (yufansong) [reply poetry.lock:1]: NVM, I already update it for you.

**Comment by JeffKatzy** [comment]
@yufansong Was just about to push, but saw that you took care of it.  Thank you for that, and for pointing it out -- it would have been tough for me to debug.

**Comment by yufansong** [comment]
> @yufansong Was just about to push, but saw that you took care of it. Thank you for that, and for pointing it out -- it would have been tough for me to debug.

Yeah, I think the previous error should be flakiness. Rerun it could solve. 
cc @xingyaoww  do you want to run eval before the merge? If not, it should be good to merge now.

**Review by yufansong — APPROVED** [review]

**Review by xingyaoww — CHANGES_REQUESTED** [review]
LGTM! I will run eval on this and merge it! "Request changes" in case it got merged by accident.

**Comment by xingyaoww** [comment]
After switch to a different machine for eval, i run into the same Python-related issue @neubig mentioned earlier -- but these's no issue if you are just interacting with `ssh_box.sh` which is pretty weird. I'll investigate it further before further running it. If https://github.com/OpenDevin/OpenDevin/pull/2722 got merged before i figure it out, i will run the eval of this PR on top of that (we just got the eval number for that!)

**Comment by xingyaoww** [comment]
@jigsawlabs-student 
Now I am able to reproduce: basically, when you run `python evaluation/swe_bench/swe_env_box.py`, then enter the command `echo 'open_file("a.txt")' | execute_cli` in the interactive shell, you will see a error:

```
l In[1], line 1                                                                                                                                               [15/1902]
----> 1 open_file("a.txt")

NameError: name 'open_file' is not defined
```

After investigating further, it seems the model failed to import aider due to a missing dependency in the `swe_env_box.py`.

image
image

I'm still investigating why the `pip install` you added in the `setup.sh` did not work..

**Comment by tobitege** [comment]
@xingyaoww the only difference I can see is the `pyproject.toml` not showing `diskcache = "5.6.3"` as that what I had in my old aider-linter PR.

**Comment by xingyaoww** [comment]
I'm still debugging my way through it... It seems the jupyter kernel is running using `/usr/bin/python3` but not the expected OpenDevin interpreter `/opendevin/miniforge3/bin/python`, which is pretty weird. I suspect this issue was introduced earlier, but because we rarely use other agentskills and the editing agentskills only need standard library, so it worked...

I'm trying to figure out how we can fix it.

image

**Comment by xingyaoww** [comment]
AND this is only an issue in the swebench eval - but not the regular sandbox we used in the unit test.. seems like a swe-bench docker specific problem.

**Comment by tobitege** [comment]
Maybe because the SWE bench script uses conda to install?
`/evaluations/swe_bench/scripts/setup/swe_entry.sh`
(IF that is applicable here)

**Comment by xingyaoww** [comment]
But supposedly, this will happen AFTER the jupyter server has been initialized..

**Comment by xingyaoww** [comment]
Dang this new commit should fixed the issue!

**Comment by JeffKatzy** [comment]
Ok, will take a look at these errors.

**Comment by xingyaoww** [comment]
@JeffKatzy I already fix the SWE-Bench errors and started the eval! I guess the remaining error to fix would be the unit tests

**Comment by JeffKatzy** [comment]
Yes, you got it.  Not sure if I will get to it by tomorrow, but otherwise handle by Sun/Monday.

**Comment by xingyaoww** [comment]
ugh, the evaluation results are not super good (59/300) despite having a similar amount of "litellm.ContentPolicyViolationError" as https://github.com/OpenDevin/OpenDevin/pull/2722#issuecomment-2221485314 this one (74/300).

image


I have uploaded the result file in this ZIP, feel free to check `output.jsonl` using the [visualizer](https://github.com/OpenDevin/OpenDevin/tree/main/evaluation/swe_bench#visualize-results):
[claude-3-5-sonnet@20240620_maxiter_30_N_v1.8-no-hint-pr2489.tar.gz](https://github.com/user-attachments/files/16200892/claude-3-5-sonnet%4020240620_maxiter_30_N_v1.8-no-hint-pr2489.tar.gz)

File: opendevin/runtime/plugins/agent_skills/aider/linter.py
Comment (xingyaoww) [inline opendevin/runtime/plugins/agent_skills/aider/linter.py:95]: Is the performance degradation because we make the link much stricter and the agent failed to fix some of them and stuck in loops?

I think the original ones are just `F821,F822,F831,E112,E113,E999,E902`

Comment (neubig) [reply opendevin/runtime/plugins/agent_skills/aider/linter.py:95]: Yeah, might be... We might actually want to have only syntax error checks by default if we want to achieve the best pass rate? 

Then we could possibly make the syntax checks configurable so people could specify the ones they want to use for their specific project

**Comment by JeffKatzy** [comment]
@xingyaoww - ok this should be passing.  I also update the `flake_lint` function so that it only runs the recommended lint checks. (ie. `fatal = 'F821,F822,F831,E112,E113,E999,E902'`).

**Comment by JeffKatzy** [comment]
@neubig @xingyaoww ok thanks graham, just re-reran `poetry lock --no-update`, which updated the `content-hash`, and then pushed.

**Comment by xingyaoww** [comment]
Did we change any packages? I got the following error when try to start the eval:

```
ERROR:root:  File "<string>", line 1, in <module>
  File "/shared/nas2/xingyao6/projects/OpenDevin-infer/agenthub/__init__.py", line 3, in <module>
    from opendevin.controller.agent import Agent
  File "/shared/nas2/xingyao6/projects/OpenDevin-infer/opendevin/controller/__init__.py", line 1, in <module>
    from .agent_controller import AgentController
  File "/shared/nas2/xingyao6/projects/OpenDevin-infer/opendevin/controller/agent_controller.py", line 5, in <module>
    from opendevin.controller.agent import Agent
  File "/shared/nas2/xingyao6/projects/OpenDevin-infer/opendevin/controller/agent.py", line 12, in <module>
    from opendevin.runtime.plugins import PluginRequirement
  File "/shared/nas2/xingyao6/projects/OpenDevin-infer/opendevin/runtime/__init__.py", line 1, in <module>
    from .docker.local_box import LocalBox
  File "/shared/nas2/xingyao6/projects/OpenDevin-infer/opendevin/runtime/docker/local_box.py", line 9, in <module>
    from opendevin.runtime.sandbox import Sandbox
  File "/shared/nas2/xingyao6/projects/OpenDevin-infer/opendevin/runtime/sandbox.py", line 7, in <module>
    from opendevin.runtime.plugins.mixin import PluginMixin
  File "/shared/nas2/xingyao6/projects/OpenDevin-infer/opendevin/runtime/plugins/__init__.py", line 2, in <module>
    from .agent_skills import AgentSkillsPlugin, AgentSkillsRequirement
  File "/shared/nas2/xingyao6/projects/OpenDevin-infer/opendevin/runtime/plugins/agent_skills/__init__.py", line 4, in <module>
    from opendevin.runtime.plugins.agent_skills.agentskills import DOCUMENTATION
  File "/shared/nas2/xingyao6/projects/OpenDevin-infer/opendevin/runtime/plugins/agent_skills/agentskills.py", line 38, in <module>
    from .aider import Linter
  File "/shared/nas2/xingyao6/projects/OpenDevin-infer/opendevin/runtime/plugins/agent_skills/aider/__init__.py", line 4, in <module>
    from .linter import Linter, LintResult
  File "/shared/nas2/xingyao6/projects/OpenDevin-infer/opendevin/runtime/plugins/agent_skills/aider/linter.py", line 9, in <module>
    from grep_ast import TreeContext, filename_to_lang

ERROR:root:<class 'ModuleNotFoundError'>: No module named 'grep_ast'
```

**Comment by JeffKatzy** [comment]
@xingyaoww Weird, looks like these packages are missing from the `pyproject.toml` file
```
grep-ast = "0.3.2"
tree-sitter = "0.21.3"
```

**Comment by opendevin-agent** [comment]
@JeffKatzy any idea how to fix the unit tests?

**Comment by xingyaoww** [comment]
@JeffKatzy any idea how to fix the unit test?

**Comment by tobitege** [comment]
> @JeffKatzy any idea how to fix the unit test?

Packages need to be re-added to pyproject.toml, merge error I guess. See tree_sitter_languages error in logs.

**Comment by JeffKatzy** [comment]
@xingyaoww Just removed the extra pytest fixture, and ruff was run again locally as part of the pre-commit hook.  Unsure if it fixed the trailing whitespace issue.

**Comment by JeffKatzy** [comment]
Is there a script to run for this?  `make lint` currently passes locally for me.

make-lint

**Comment by JeffKatzy** [comment]
@xingyaoww ok I ran, `pre-commit run --from-ref origin/main --to-ref HEAD` locally and it updated a couple files.  Hopefully this fixed it.

**Comment by JeffKatzy** [comment]
@xingyaoww ok I ran, `pre-commit run --from-ref origin/main --to-ref HEAD` locally and it updated a couple files.  Hopefully this fixed it.

**Comment by tobitege** [comment]
> @xingyaoww ok I ran, `pre-commit run --from-ref origin/main --to-ref HEAD` locally and it updated a couple files. Hopefully this fixed it.

Latest commit looks clean to me.

**Comment by JeffKatzy** [comment]
@xingyaoww @tobitege Hope the above works.  I'm blocked from doing a rebase on main as the latest refactor of the config results in failing tests and am unable to resolve.

**Comment by JeffKatzy** [comment]
@xingyaoww can you run the CI/CD when you have a chance. Previous run was failing because of the poetry.lock file.

**Comment by xingyaoww** [comment]
Sorry for the long wait! This has been taking forever...
78/300=26% - a small boost! Let's go ahead and merge it!
image

**Comment by tobitege** [comment]
That's good news!  Errr... you gotta switch your review for merge, though 😁

**Review by xingyaoww — APPROVED** [review]
Good to go 🚀

**Pull request merged**
